### PR TITLE
big, breaking, rewrite

### PR DESCRIPTION
**Pull request by tjtnew**
Large, breaking, rewrite of report factory.  For initial CRAN release the tasks left to do are:
* rework the old vignettes;
* update artwork;
* ensure tests are encompassing.
* Add some interactive clean_reports functionality.

### CODE REVIEW

File: R/compile_reports.R [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    # pull yaml from the report
+    yaml <- rmarkdown::yaml_front_matter(r)
+
+    # If params are supplied, these are combined with any default parameters
+    # that may be set in the report header.  Where there are overlaps preference
+    # is given to those set here.  After much trial and error the current way
+    # we facilitate this is to alter the yaml header and save this altered
+    # version to a new file which we then compile.
+    if (!is.null(params)) {
```
Comment (thibautjombart) [inline R/compile_reports.R:67]: Given the constraints and multiplte caveats we've seen on this, I think the current solution is actually pretty cool. Who knows, maybe `render` will eventually default to clean environments, in which case we can simplify the code. This said, I really like the trick of changing the yaml. Brutal, but fun!

File: R/factory_root.R [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#'
+#' @param directory the full path to the directory you are interested in.  If
+#'   NULL (default) the function assumes you are already in a factory and will
+#'   return the address of the containing factory root.
+#'
+#' @noRd
+#' @keywords internal
+factory_root <- function(directory = NULL) {
+  
+  if(!is.null(directory)) {
```
Comment (thibautjombart) [inline R/factory_root.R:13]: I'm probably thick here, but would it be simpler to start off with something like:
```

if (is.null(directory)) {
  directory <- getwd()
}
```
or even define `directory = getwd()` in the arguments?

File: R/install_deps.R [hunk 3]
```diff
@@ -1,29 +1,24 @@
...
-
-install_deps <- function(update = FALSE, ...) {
+#' 
+#' @export
+install_deps <- function(factory = ".", update = FALSE, ...) {
 
-  pkg_to_install <- list_deps(!update)
+  pkg_to_install <- list_deps(factory = factory, missing = !update)
 
   if (length(pkg_to_install) > 0L) {
```
Comment (thibautjombart) [inline R/install_deps.R:19]: I tend to use `if (length(xxx))` nowadays, but cosmetic; just writing this to know if there's an emerging consensus on what's best.

File: R/internals.R [hunk 4]
```diff
@@ -0,0 +1,121 @@
+#' Find the path to the root folder within a factory
+#'
+#' This function is used to find the fully expanded root path within a factory.
+#'
+#' @param directory the full path to the directory you are interested in.  If
+#'   NULL (default) the function assumes you are already in a factory and will
+#'   return the address of the containing factory root.
+#'
+#' @noRd
+factory_root <- function(directory = NULL) {
```
Comment (thibautjombart) [inline R/internals.R:10]: Same comment as elsewhere (can't remember where, sorry); do we want to default to 
`directory = "."` or `directory = getwd()` here and remove the leading `if` statement?

File: R/new_factory.R [hunk 5]
```diff
@@ -4,86 +4,132 @@
...
+#'   created; defaults to TRUE.
+#' @param create_example_report A `logical` indicating if `new_factory()` should
+#'   create an example report in the 'report_sources' folder along with some
+#'   example data in the 'data/raw' folder; defaults to TRUE.
+#' @param create_data_folders a `logical` indicating if `new_factory()` should
+#'   create folders to store data; defaults to TRUE.
+#' @param create_scripts_folder a `logical` indicating if `new_factory()` should
+#'   create folders to store R scripts; defaults to TRUE.
+#' @param use_here a `logical` indicating if `new_factory()` should create
+#'   a `.here` that can be used with `here::here()`.
```
Comment (thibautjombart) [inline R/new_factory.R:26]: Add: 
Defaults to `TRUE`

File: R/rfh_load_scripts.R [hunk 6]
```diff
@@ -1,92 +0,0 @@
-#' Load all R scripts in the factory
-#'
-#' This helper function will load all `.R` scripts within folders `scripts/` and
```
Comment (thibautjombart) [inline R/rfh_load_scripts.R:3]: Can we re-add this file with its test without damaging anything?

File: R/validate_factory.R [hunk 7]
```diff
@@ -1,143 +1,134 @@
...
-#'
-#' }
-#'
+#' Checks run on the factory include:
+#'   * the factory directory exists;
+#'   * the factory_config file exist;
+#'   * all mandatory folders exist - by default these are 'report_sources/'
+#'     and 'outputs/';
+#'   * (optionally) all .Rmd reports have unique names once outside their
+#'     folders.
```
Comment (thibautjombart) [inline R/validate_factory.R:18]: Putting a pin in this for discussing, but also to chek. Assuming we have:
report_sources/foo.Rmd
report_sources/bar/foo.Rmd
Then, is the the case that:
`compile_reports("foo")`
will compile both but for any duplicated output (i.e. output with the same name), only the second version will be stored in `outputs/foo/[timestamp]`?
Or will it generate a mirrored output folder structure? If the latter is true, then we may no longer want to check that reports have unique names outside their folders.

File: R/validate_factory.R [hunk 8]
```diff
@@ -1,143 +1,134 @@
...
-      msg <- sprintf("the directory '%s' does not exist", factory)
-      stop(msg)
+validate_factory <- function(factory = ".", allow_duplicates = TRUE) {
+
+  # this finds the folder containing factory_config or errors if not possible
+  root <- factory_root(factory)
+
+  # load configuration
+  config <- as.data.frame(read.dcf(fs::path(root, "factory_config")))
+
```
Comment (thibautjombart) [inline R/validate_factory.R:34]: Petty comment again, but would it make sense to externalise this? Trying to think if we'll repeatedly need to check the config, or if we'll want to show it to the user. Note, we can always do this later.

File: README.Rmd [hunk 9]
```diff
@@ -5,36 +5,47 @@ output: github_document
...
-knitr::opts_knit$set(root.dir = tempdir())
 ```
 
-[![Travis-CI Build Status](https://travis-ci.org/reconhub/reportfactory.svg?branch=master)](https://travis-ci.org/reconhub/reportfactory)
-[![Build status](https://ci.appveyor.com/api/projects/status/7h2mgej230dv5r7w/branch/master?svg=true)](https://ci.appveyor.com/project/thibautjombart/reportfactory/branch/master)
-[![Coverage Status](https://codecov.io/github/reconhub/reportfactory/coverage.svg?branch=master)](https://codecov.io/github/reconhub/reportfactory?branch=master)
-[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/reportfactory)](https://cran.r-project.org/package=reportfactory)
-[![CRAN Downloads](https://cranlogs.r-pkg.org/badges/reportfactory)](https://cran.r-project.org/package=reportfactory)
+<!-- badges: start -->
+[![R-CMD-check](https://github.com/reconhub/reportfactory/workflows/R-CMD-check/badge.svg)](https://github.com/reconhub/reportfactory/actions)
```
Comment (thibautjombart) [inline README.Rmd:17]: I love that we are now using github actions for checks :)

File: README.Rmd [hunk 10]
```diff
@@ -47,72 +58,82 @@ To install the development version of the package, use:
...
+passing full filenames to the function so it is always clear what will be built.
 
-Use helper functions to see progress.
-```{r compile_one_results}
-list_outputs(factory_proj)
+```{r compile}
+compile_reports( 
+  reports = "example_report.Rmd"
+)
 ```
```
Comment (thibautjombart) [inline README.Rmd:107]: Suggest adding:
Note that `reports` can also be an `integer` or a `logical` vector, in which case it is interpreted as a subset of files output by `list_reports()`. For instance:
* `reports = c(1, 3)` will compile the first and third reports listed by `list_reports()`
* `reports = TRUE` will compile all reports

File: README.Rmd [hunk 11]
```diff
@@ -47,72 +58,82 @@ To install the development version of the package, use:
...
-### Step 4 - Consolidate latest outputs
-You can now get a single directory of the latest report outputs using:
-
-```{r shipit}
-ship_reports(factory = factory_proj, most_recent = TRUE)
-list.dirs(factory_proj)
+If you want to have an overview of your entire factory then you can use the
+`fs` package and the `dir_tree` function:
+```{r dirtree}
+fs::dir_tree()
```
Comment (thibautjombart) [inline README.Rmd:129]: It is a bit cheeky, but would we want a wrapper e.g. `factory_overview()` for this?

File: inst/skeletons/README.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+## How the factory is organised
+
+  - `report_sources`: (mandatory) put your `.Rmd` documents there
+    (subfolders are OK)
+
+  - `data/`: (recommended) put your data in this folder (subfolders are
+    OK)
+    
+      - `data/raw`: to store raw data, as read-only
```
Comment (thibautjombart) [inline inst/skeletons/README.md:15]: I think this file needs updating; check that the default structure outlined is correct, and update with new functions.

File: inst/skeletons/README.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  - `list_reports()`: lists reports currently stored in the factory
+    (only `.Rmd` source files)
+
+  - `compile_report()`: compiles one single report, designated using the
+    file name or a non-ambiguous match; outputs will be stored in
+    `outputs/`
+
+  - `update_reports()`: compiles all report, using by default the most
+    recent version of each report; outputs will be stored in
+    `outputs/`
```
Comment (thibautjombart) [inline inst/skeletons/README.md:35]: Remove

File: inst/skeletons/README.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+## How to run the factory: useful commands
+
+  - `list_reports()`: lists reports currently stored in the factory
+    (only `.Rmd` source files)
+
+  - `compile_report()`: compiles one single report, designated using the
+    file name or a non-ambiguous match; outputs will be stored in
+    `outputs/`
+
```
Comment (thibautjombart) [inline inst/skeletons/README.md:32]: plural, update description

File: inst/skeletons/README.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+3.  check your report by compiling the `.Rmd` manually if needed,
+    e.g. `rmarkdown::render("foobar.Rmd")`; once you are
+    happy with the results, **make sure you remove all output files from
+    the source folder**
+
+4.  run `update_reports()` to generate all outputs, or
+    `compile_report("foobar")` if you just want to produce
+    time-stamped outputs for this report; check results in the folder
+    `outputs`
```
Comment (thibautjombart) [inline inst/skeletons/README.md:53]: update this

File: inst/skeletons/example_report.Rmd [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+author: "Joe Bloggs"
+params:
+  grouped_plot: TRUE
+output:
+  html_document: default
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, collapse = TRUE, fig.width = 8, fig.height = 6, dpi = 150)
```
Comment (thibautjombart) [inline inst/skeletons/example_report.Rmd:12]: Petty, but this makes for large figures to start with.`dpi = 70` should be good enough

File: tests/testthat/test-compile_report.R [hunk 17]
```diff
@@ -0,0 +1,175 @@
+library(fs)
+
+test_that("test parameteriesed report output", {
```
Comment (thibautjombart) [inline tests/testthat/test-compile_report.R:3]: typo; should this file be named `test-compile_reports`?

**Review by thibautjombart — COMMENTED** [review]
Lots of comments, mostly petty things you should feel free to ignore, and a few things I wanted to put pins in for discussion / clarification (mostly to make sure I know what expected behaviours are). Overall, the rewrite is awesome, and the code is very clear (which I can't emphasize enough how cool this is!). 

If possible, consider re-adding `rfh_load_scripts()` and its test. Happy to do some re-write if needed. I suspect the tests will need tweaking, to adopt the current design of tests, replacing `random_factory()` with `new_factory(fs::path_temp())` followed by removal on exit.

File: .github/ISSUE_TEMPLATE/issue_template.md [hunk 18]
```diff
@@ -1,19 +0,0 @@
...
---------
-
-Please include a brief description of the problem with a code example:
-
-```r
-# insert reprex here
-```
-
----------
-
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply .github/ISSUE_TEMPLATE/issue_template.md:19]: Agree - will probably end up including one for all recon repos. Going to think about it a little more first and apply across the board.

File: .github/PULL_REQUEST_TEMPLATE/pull_request_template.md [hunk 19]
```diff
@@ -1,25 +0,0 @@
-* **Please check if the PR fulfills these requirements**
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:1]: same response as template for issues response

File: DESCRIPTION [hunk 20]
```diff
@@ -1,23 +1,32 @@
...
-Description: Provides an infrastructure for handling multiple dated reports using \code{rmarkdown::render}, including automated curation and time-stamping of outputs, parametrisation, dependencies, and some helper functions.
+Version: 0.0.7
+Authors@R: c(
+  person("Thibaut", "Jombart", role = c("aut", "cre"), email = "thibautjombart@gmail.com"), 
+  person("Zhian N.", "Kamvar", role = "aut", email = "zkamvar@gmail.com"),
+  person("Amy", "Gimma", role = "ctb", email = "amyg225@gmail.com"),
+  person("Tim", "Taylor", role ="aut", email = "tim.taylor@hiddenelephants.co.uk",
+         comment = c(ORCID = "0000-0002-8587-7113"))
+ )
+Description: Provides an infrastructure for handling multiple dated reports
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply DESCRIPTION]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 21]
```diff
@@ -0,0 +1,159 @@
+#' Compile an R Markdown report
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R]: https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+#' 
+#' @param factory The path to the report factory or a folder within the desired
+#'   factory. Defaults to the current directory.
+#' @param reports Either a regular expression (passed directly to `grep()`) that
+#'   matches to the report paths you would like to compile or an integer/logical
+#'   vector.  If `reports` is an integer or character vector then a call of
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R]: https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 23]
```diff
@@ -0,0 +1,159 @@
+#' Compile an R Markdown report
+#' 
+#' @param factory The path to the report factory or a folder within the desired
+#'   factory. Defaults to the current directory.
+#' @param reports Either a regular expression (passed directly to `grep()`) that
+#'   matches to the report paths you would like to compile or an integer/logical
+#'   vector.  If `reports` is an integer or character vector then a call of 
+#'   `compile_reports(factory, reports = idx)` is equivalent to
+#'   `compile_reports(factory, list_reports(factory)[idx])`.
+#' @param params A named list of parameters to be fed to each report that
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+#'   matches the regular expression.  If a report has default parameters set in
+#'   it's yaml heading and their are duplicates then preference is given to the
+#'   value supplied here. 
+#' @param quiet A logical indicating if messages from R Markdown compilation
+#'   should be displayed; `TRUE` by default.
+#' @param timestamp timevalue used in the folder structure of the report output.
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+#'   If NULL, `format(Sys.time(), "%Y-%m-%d_T%H-%M-%S")` will be used and it's
+#'   evaluation forced immediately.
+#' @param subfolder Name of subfolder to store results.  Not required but helps
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R:23]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/factory_root.R (hunk 2 above)
Comment (tjtnew) [reply R/factory_root.R]: Yes - no idea why I did this.  Using `"."` over getwd() as it seems like one less function call.  closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/install_deps.R [hunk 26]
```diff
@@ -1,29 +1,24 @@
...
-#'   should be re-installed (\code{TRUE}); otherwise, only missing packages are
-#'   installed; defaults to \code{FALSE}.
-#'
-#' @param ... Arguments to be passed to \code{install.packages}.
-#'
+#'   should be re-installed (`TRUE`); otherwise, only missing packages are
+#'   installed; defaults to `FALSE`.
+#' @param ... Arguments to be passed to `install.packages()`.
+#' 
 #' @seealso \code{\link{list_deps}} to list dependencies of packages
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/install_deps.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/install_deps.R (hunk 3 above)
Comment (tjtnew) [reply R/install_deps.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/internals.R (hunk 4 above)
Comment (tjtnew) [reply R/internals.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/new_factory.R [hunk 27]
```diff
@@ -4,86 +4,132 @@
...
-#' outputs in \code{report_outputs}
-#'
-#'  \item \code{open.Rproj}: an Rproject to open the factory using Rstudio
+#' @param factory The name of the report factory to be created.
+#' @param path The folder where the report factory should be created.  This
+#'   will default to the current directory.
+#' @param report_sources The name of the folder to be used for report
+#'   templates.  Defaults to 'report_sources/'.
+#' @param outputs The name of the folder to be used for saving the built
+#'   reports.  Defaults to 'outputs/'.
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/new_factory.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/new_factory.R (hunk 5 above)
Comment (tjtnew) [reply R/new_factory.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: inst/skeletons/README.md (hunk 13 above)
Comment (tjtnew) [reply inst/skeletons/README.md]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: inst/skeletons/example_report.Rmd (hunk 16 above)
Comment (tjtnew) [reply inst/skeletons/example_report.Rmd]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: tests/testthat/test-compile_report.R (hunk 17 above)
Comment (tjtnew) [reply tests/testthat/test-compile_report.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/list_outputs.R [hunk 28]
```diff
@@ -1,28 +1,32 @@
-
 #' List outputs of the factory
 #'
 #' This function can be used to list available report outputs stored in the
-#' factory, in inside the \code{report_outputs} folder (or any subfolder
+#' factory, in inside the \code{outputs} folder (or any subfolder
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/list_outputs.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/e60198f05faaf899a33aab96f3d9634ab3413252

File: R/compile_reports.R [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+compile_reports <- function(factory = ".", reports = NULL,
+                            params = NULL, quiet = TRUE, subfolder = NULL,
+                            timestamp = format(Sys.time(), "%Y-%m-%d_T%H-%M-%S"), 
+                            ...) {
+  
+  # force timestamp to evaluate as soon as function called - needed due to the
+  # `Sys.time` call within the default argument.
+  force(timestamp)
+  
+  # move in to factory
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply R/compile_reports.R]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/640ce7826ac4440fe95cf7249966d8c02a2fc389

File: inst/skeletons/README.md (hunk 14 above)
Comment (tjtnew) [reply inst/skeletons/README.md]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/640ce7826ac4440fe95cf7249966d8c02a2fc389

File: NAMESPACE [hunk 30]
```diff
@@ -1,14 +1,10 @@
 # Generated by roxygen2: do not edit by hand
 
-export(archive_reports)
-export(clean_report_sources)
-export(compile_report)
+export(compile_reports)
```
[the remark this answers is not in the document]
Comment (tjtnew) [reply NAMESPACE:3]: closed in https://github.com/reconhub/reportfactory/pull/68/commits/aebd94971e06de3738dc2bd11a316cf142aebca9

File: R/rfh_load_scripts.R (hunk 6 above)
Comment (tjtnew) [reply R/rfh_load_scripts.R:3]: closed in https://github.com/reconhub/reportfactory/pull/68/commits/aebd94971e06de3738dc2bd11a316cf142aebca9

File: R/validate_factory.R (hunk 7 above)
Comment (tjtnew) [reply R/validate_factory.R]: The latter is correct.  I've removed the duplicates check in https://github.com/reconhub/reportfactory/pull/68/commits/c6bba8180dba8e6a811501d54784583c36f5bc54

File: README.Rmd (hunk 10 above)
Comment (tjtnew) [reply README.Rmd:107]: closed with https://github.com/reconhub/reportfactory/pull/68/commits/8a9da6afe3f55c2610c00e801e0dfb8830e70b4e

File: inst/skeletons/README.md (hunk 12 above)
Comment (tjtnew) [reply inst/skeletons/README.md]: to be checked but hopefully closed with https://github.com/reconhub/reportfactory/pull/68/commits/c022fc73b1360e4806282a532c3079af6912575f

File: inst/skeletons/README.md (hunk 15 above)
Comment (tjtnew) [reply inst/skeletons/README.md]: To be checked but hopefully closed with https://github.com/reconhub/reportfactory/pull/68/commits/c022fc73b1360e4806282a532c3079af6912575f

**Comment by tjtnew** [comment]
@thibautjombart  A couple of comments to resolve once we talk.  Can you look over:
* the updated README.
* the updated skeleton README.
* the load_scripts.R file

File: R/validate_factory.R (hunk 8 above)
Comment (tjtnew) [reply R/validate_factory.R:30]: Let's think about this in the follow-up release.

File: README.Rmd (hunk 11 above)
Comment (tjtnew) [reply README.Rmd]: added in https://github.com/reconhub/reportfactory/pull/68/commits/63d2ad10489e459cfff67572a14f15491d0b059b

**Review by thibautjombart — APPROVED** [review]
:)

**Pull request merged**